pkg/services/ngalert/remote/client/alertmanager.go

```diff
@@ -0,0 +1,113 @@
+package client
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"time"
+
+	httptransport "github.com/go-openapi/runtime/client"
+	"github.com/grafana/grafana/pkg/infra/log"
+	amclient "github.com/prometheus/alertmanager/api/v2/client"
+)
+
+const alertmanagerAPIMountPath = "/alertmanager"
+const alertmanagerReadyPath = "/-/ready"
+
+type AlertmanagerConfig struct {
+	TenantID string
+	Password string
+	URL      *url.URL
+	Logger   log.Logger
+}
+
+type Alertmanager struct {
+	*amclient.AlertmanagerAPI
+	httpClient *http.Client
+	url        *url.URL
+	logger     log.Logger
+}
+
+func NewAlertmanager(cfg *AlertmanagerConfig) (*Alertmanager, error) {
+	// First, add the authentication middleware.
+	c := &http.Client{Transport: &MimirAuthRoundTripper{
+		TenantID: cfg.TenantID,
+		Password: cfg.Password,
+		Next:     http.DefaultTransport,
+	}}
+
+	apiEndpoint := *cfg.URL
+
+	// Next, make sure you set the right path.
+	u := apiEndpoint.JoinPath(alertmanagerAPIMountPath, amclient.DefaultBasePath)
+	transport := httptransport.NewWithClient(u.Host, u.Path, []string{u.Scheme}, c)
+
+	return &Alertmanager{
+		logger:          cfg.Logger,
+		url:             cfg.URL,
+		AlertmanagerAPI: amclient.New(transport, nil),
+		httpClient:      c,
+	}, nil
+}
+
+// GetAuthedClient returns a *http.Client that includes a configured MimirAuthRoundTripper.
+// Requests using this client are fully authenticated.
+func (am *Alertmanager) GetAuthedClient() *http.Client {
+	return am.httpClient
+}
+
+// IsReadyWithBackoff executes a readiness check against the `/-/ready` Alertmanager endpoint.
+// If it takes more than 10s to get a response back - we abort the check.
+func (am *Alertmanager) IsReadyWithBackoff(ctx context.Context) (bool, error) {
+	ctx, cancel := context.WithCancel(ctx)
```

Comment:
We can skip creating the timer below.
```suggestion
	ctx, cancel := context.WithTimeout(ctx, 10 * time.Second)
```

Reply:
Ditto from above.